sk-gen/src/main.rs

```diff
@@ -0,0 +1,700 @@
+#![deny(rustdoc::broken_intra_doc_links)]
+//! `sk-gen` is a CLI tool for generating synthetic trace data for SimKube.
+//!
+//! # Overview:
+//! ## Core types
+//! [`Node`] represents a cluster state, containing a map from unique names to active
+//! [`Deployment`] states. `Node` implements [`Eq`] and [`Hash`], which we use to ensure that
+//! equivalent `Node`s are not duplicated in the graph.
+//!
+//! [`Deployment`] is a simplified representation of a Kubernetes deployment spec, containing only
+//! the fields we are considering.
+//!
+//! [`DeploymentAction`] (e.g. `CreateDeployment`, `DeleteDeployment`, `IncrementReplicas`,
+//! `DecrementReplicas`) can be performed on individual deployment instances.
+//!
+//! [`ClusterAction`] contains a name of a candidate deployment alongside a [`DeploymentAction`]
+//! such that it can be applied to a `Node` without ambiguity as to which deployment it applies. Not
+//! all `DeploymentAction`s are valid for every `Deployment`, and neither are all `ClusterAction`
+//! instances valid for every `Node`. For instance, we cannot delete a `Deployment` that does not
+//! exist, nor can we increment/decrement the replicas of a `Deployment` that is not active.
+//!
+//! [`TraceEvent`] represents the Kubernetes API call which corresponds to a `ClusterAction`.
+//!
+//! [`Edge`] stores both a `ClusterAction` and the corresponding `TraceEvent`.
```

Comment:
Why is ClusterAction not a property of TraceEvent?

Reply:
We don't have control over TraceEvent, which is the atomic unit of change ingested by simkube. ClusterAction is internal to our generation, we cache the TraceEvent on the edge so that we don't have to regenerate every walk. Other approaches may be better suited to that task.